src/lib/scheduler/node_queue_scheduler.cpp

```diff
@@ -76,22 +82,46 @@ void NodeQueueScheduler::wait_for_all_tasks() {
       // concurrent queues when many tiny tasks have been scheduled (see MergeSort scheduler test). When this assert is
       // triggered in other situations, there have probably been new tasks added after wait_for_all_tasks() was called.
       Assert(queue_check_runs < 1'000, "Queue is not empty but all registered tasks have already been processed.");
+
+      // Explicitly signal workers to work on remaining jobs.
+      // TODO(Martin): still needed?
+      // queue->signal(1);
+
       std::this_thread::sleep_for(std::chrono::milliseconds(1));
       ++queue_check_runs;
     }
   }
 }
 
 void NodeQueueScheduler::finish() {
+  // Lock finish() to ensure that the shutdown tasks are not sent twice.
+  auto lock = std::lock_guard<std::mutex>{_finish_mutex};
+
+  if (!_active) {
+    return;
+  }
+
   wait_for_all_tasks();
 
-  // All queues SHOULD be empty by now
-  if (HYRISE_DEBUG) {
-    for (auto& queue : _queues) {
-      Assert(queue->empty(), "NodeQueueScheduler bug: Queue wasn't empty even though all tasks finished");
+  auto workers_shut_down = std::atomic_uint64_t{_worker_count};
+
+  for (auto node_id = NodeID{0}; node_id < _node_count; ++node_id) {
+    const auto node_worker_count = _workers_per_node[node_id];
+    for (auto worker_id = size_t{0}; worker_id < node_worker_count; ++worker_id) {
+      // Create a shutdown task for every worker.
+      auto job_task = std::make_shared<JobTask>([&]() { --workers_shut_down; }, SchedulePriority::Default, false);
+      job_task->set_as_shutdown_task();
```

Comment:
I wonder if we should make this a bit more explicit than having a JobTask with a flag. I think distributed systems, e.g., Akka, have a dedicated message/task for this. I'm fine with the `is_shutdown_task()` function in `AbstractTask` to avoid a dynamic cast, but the flag could be set to false for all but special `ShutdownTask`s (thus, also avoiding a virtual function call if just one always returned false and the other always returned true). You could even simplify the interface of such a `ShutdownTask` to only take a reference to the variable/the scheduler then (or the `ShutdownTask` gets the scheduler from the Hyrise singleton). This should also ease adding a test that checks if the scheduler actually sends such a task to each worker.

Reply:
Goof point with the static variable. I decided against the own class mainly due to the ugly dynamic cast (there should be a commit implementing that somewhere in the history).

What would be the purpose of the reference to the scheduler?

Reply:
I just brainstormed some ideas how a `ShutdownTask` could look like. It should not take a lambda with the variable decrement as an init parameter and could just get a ref to the scheduler or the counter variable to decrement the counter on execution. Just some ideas though. I think handing something to the task eases testing.